server/oauth.js

```diff
@@ -0,0 +1,170 @@
+				+ 'scope=MiEmpresa&'
+				+ 'redirect_uri=' + encodeURIComponent(env.oauth.redirectUrl)
+		});
+		res.end();
+	},
+	callbackMiddleware: function (req, res, next) {
+		var query = req.query;
+
+		if (req._parsedUrl.pathname !== '/eid-callback') {
+			next();
+			return;
+		}
+
+		debug('Callback received. Query:', query);
+
+		if (query.error) {
+			// TODO: How should we notify the user about the error? Flash message?
+			res.writeHead(302, { Location: '/' });
+			res.end();
+
+			return;
+		}
+
+		var state = req.cookies.get('OAuthState');
+
+		// Clear state as it's supposed to be one time only.
+		res.cookies.set('OAuthState', null);
+
+		if (state !== query.state) {
+			debug('State value did not match:', state, query.state);
+
+			// TODO: Notify user.
+			res.writeHead(302, { Location: '/' });
+			res.end();
+
+			return;
+		}
+
+		debug('States match');
+
+		request({
+			uri: env.oauth.tokenEndpoint,
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded',
+				Authorization: 'Basic ' + new Buffer(env.oauth.clientId + ':'
+					+ env.oauth.clientSecret).toString('base64')
+			},
+			form: {
+				grant_type: 'authorization_code',
+				code: query.code,
+				redirect_uri: env.oauth.redirectUrl
+			}
+		}, function (error, response, body) {
+			if (error) {
+				debug('Error received from token endpoint:', error);
+
+				// TODO: Notify user.
+				res.writeHead(302, { Location: '/' });
+				res.end();
+			} else if (response.statusCode >= 200 && response.statusCode < 300) {
+				var parsedBody  = JSON.parse(body)
+				  , accessToken = parsedBody.access_token
+				  , decoded     = jwtDecode(accessToken);
+
+				debug('JWT received for:', decoded.email);
+
+				userEmailMap(function (map) {
+					return map.get(serializeValue(decoded.email));
+				})(function (userId) {
+					if (userId) return userId;
+
+					var isPublicApp = req.$appName === 'public'
+					  , demoUserId  = isPublicApp ? null : res.cookies.get('demoUser')
+					  , records     = [];
+
+					if (!demoUserId) {
+						return mano.queryMemoryDb([], 'addUser', JSON.stringify({
+							firstName: decoded.fname,
+							lastName: decoded.lname,
+							email: decoded.email,
+							roles: ['user']
+						}));
+					}
+
+					userId = demoUserId;
+
+					records.push(dbjsDataRecord(userId + '/isDemo', undefined));
+					records.push(dbjsDataRecord(userId + '/firstName', decoded.fname));
+					records.push(dbjsDataRecord(userId + '/lastName', decoded.lname));
+					records.push(dbjsDataRecord(userId + '/email', decoded.email));
+
+					return userStorage.storeMany(records)(function () {
+						return loadToMemoryDb([userId]);
+					})(function () {
+						return userId;
+					});
+				}).done(function (userId) {
+					if (!decoded.email_verified) {
+						res.writeHead(302, { Location: '/request-confirm-account/' });
+						res.end();
+						return;
+					}
+
+					login(userId, req, res);
+
```

Comment:
Let's pretty please ensure we have some handle on the token to issue further requests, like: https://github.com/egovernment/eregistrations-salvador/pull/2071/files#diff-fb19faa00022c30c2cb25d27383803fcR121